tests/test_class2D.py

```diff
@@ -313,6 +320,74 @@ def testImplementations(self):
         ):
             _ = RIRClass2D(self.clean_src, self.basis)
 
+    def testSelectionImplementations(self):
+        """
+        Test optional implementations handle bad inputs with a descriptive error.
+        """
+
+        class CustomClassSelector(ClassSelector):
+            def __init__(self, x):
+                self.x = x
+
+            def _select(self, n, classes, reflections, distances):
+                return self.x
+
+        # lower bound
+        with pytest.raises(ValueError, match=r".*out of bounds.*"):
+            rir = RIRClass2D(
+                self.clean_src,
+                self.clean_fspca_basis,
+                n_classes=self.n_classes,
+                selector=CustomClassSelector(np.arange(self.n_classes) - 1),
+            )
+            _ = rir.averages(*rir.classify())
+
+        # upper bound
+        with pytest.raises(ValueError, match=r".*out of bounds.*"):
+            rir = RIRClass2D(
+                self.clean_src,
+                self.clean_fspca_basis,
+                n_classes=self.n_classes,
+                selector=CustomClassSelector(
+                    np.arange(self.n_classes) + self.clean_src.n
+                ),
+            )
+            _ = rir.averages(*rir.classify())
+
+        # empty
+        with pytest.raises(ValueError, match=r".*must be len.*"):
+            rir = RIRClass2D(
+                self.clean_src,
+                self.clean_fspca_basis,
+                n_classes=self.n_classes,
+                selector=CustomClassSelector([]),
+            )
+            _ = rir.averages(*rir.classify())
```

Comment:
I guess "empty" covers "too short", right? Would there be any reason to test both?